src/audio/drc/drc.c

```diff
@@ -269,6 +269,43 @@ static int drc_process(struct processing_module *mod,
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
+static int drc_params(struct processing_module *mod)
+{
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct sof_ipc_stream_params comp_params;
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *sinkb;
+	struct comp_buffer __sparse_cache *sink_c;
+	enum sof_ipc_frame valid_fmt, frame_fmt;
+	int i, ret;
+
+	comp_dbg(dev, "drc_params()");
+
+	comp_params = *params;
+	comp_params.channels = mod->priv.cfg.base_cfg.audio_fmt.channels_count;
+	comp_params.rate = mod->priv.cfg.base_cfg.audio_fmt.sampling_frequency;
+	comp_params.buffer_fmt = mod->priv.cfg.base_cfg.audio_fmt.interleaving_style;
+
+	audio_stream_fmt_conversion(mod->priv.cfg.base_cfg.audio_fmt.depth,
+				    mod->priv.cfg.base_cfg.audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
```

Comment:
don't know what this function does, but seems weird we are throwing away `valid_fmt`

Reply:
please test before submitting

Reply:
@singalsu any word on this?

Reply:
Sorry, not yet. We had heavy focus in code optimization in last weeks. I Hope I can do this next week.

Reply:
I made this topology as sof-tgl-nocodec.tplg replacement for UPX device. Drc.0.1 is s16 input/output, similarly drc.1.1 is s24, drc2.1 is s32. The playback PCMs accept all s16/s24/s32 formats. The I2S DAIs are s32. The capture PCMs get SSP loopback audio.

![sof-tgl-nocodec-drc-formats](https://github.com/thesofproject/sof/assets/28754793/2ecb734c-c0cf-45c2-84cb-1acae4cd8145)

I'll make tomorrow script to test all formats combinations.

Reply:
@cujomalainey Now confirmed, the next line `comp_params.frame_fmt = frame_fmt;` needs to be `comp_params.frame_fmt = valid_fmt;`. Good catch! And need to fix this from other components too. It causes s24_le format to be treated as s32_le.

Reply:
Glad it was caught, thanks for testing :)